libs/src/api/user.js

```diff
@@ -458,12 +506,100 @@ class Users extends Base {
+    this.IS_OBJECT(newMetadata)
+    const phases = {
+      UPDATE_CONTENT_NODE_ENDPOINT_ON_CHAIN: 'UPDATE_CONTENT_NODE_ENDPOINT_ON_CHAIN',
+      UPLOAD_METADATA: 'UPLOAD_METADATA',
+      UPDATE_METADATA_ON_CHAIN: 'UPDATE_METADATA_ON_CHAIN',
+      ASSOCIATE_USER: 'ASSOCIATE_USER'
+    }
+    let phase = ''
+
+    const oldMetadata = this.userStateManager.getCurrentUser()
+    if (!oldMetadata) { throw new Error('No current user.') }
+
+    newMetadata = this._cleanUserMetadata(newMetadata)
+    this._validateUserMetadata(newMetadata)
+
+    // If the new metadata is the same as the original, it is a no-op and exit early
+    // NOTE: be careful -- this implies that the userStateManager is the source of truth
+    // if userStateManager is not up to date, can cause issues
+    const metadataFields = this._getMetadataFieldsToUpdate(newMetadata)
+    if (metadataFields.include.length === 0) { return }
+
+    try {
+      // Update user creator_node_endpoint on chain if applicable
+      if (newMetadata.creator_node_endpoint !== oldMetadata.creator_node_endpoint) {
+        phase = phases.UPDATE_CONTENT_NODE_ENDPOINT_ON_CHAIN
+        await this.contracts.UserFactoryClient.updateCreatorNodeEndpoint(userId, newMetadata['creator_node_endpoint'])
+        // Ensure DN has indexed creator_node_endpoint change
+        await this._waitForCreatorNodeEndpointIndexing(userId, newMetadata.creator_node_endpoint)
+      }
+
+      // Upload new metadata object to CN
+      phase = phases.UPLOAD_METADATA
+      const { metadataMultihash, metadataFileUUID } = await this.creatorNode.uploadCreatorContent(newMetadata)
+
+      // Write metadata multihash to chain
+      phase = phases.UPDATE_METADATA_ON_CHAIN
+      const updatedMultihashDecoded = Utils.decodeMultihash(metadataMultihash)
+      const { txReceipt } = await this.contracts.UserFactoryClient.updateMultihash(userId, updatedMultihashDecoded.digest)
+
+      // Write remaining metadata fields to chain
+      const { latestBlockNumber } = await this._updateUserOperations(newMetadata, oldMetadata, userId, metadataFields.exclude)
+
+      // Write to CN to associate blockchain user id with updated metadata and block number
+      phase = phases.ASSOCIATE_USER
+      await this.creatorNode.associateCreator(userId, metadataFileUUID, Math.max(txReceipt.blockNumber, latestBlockNumber))
+
+      // Update libs instance with new user metadata object
+      this.userStateManager.setCurrentUser({ ...oldMetadata, ...newMetadata })
+    } catch (e) {
+      // TODO: think about handling the update metadata on chain and associating..
+      throw new Error(`updateAndUploadMetadata() Error -- Phase ${phase}: ${e}`)
+    }
+  }
+
+  /**
+  * Compares the original user state to an updated state. Determines the fields to include or exclude
+  * in the metadata update on chain.
+  * @param {Object} newMetadata fields to update in the current user state
+  */
+  _getMetadataFieldsToUpdate (newMetadata) {
```

Comment:
also we need to change `newMetadata` to not be passed by reference

Reply:
going to remove this method